docs/proposals/ose-rebase.md

```diff
@@ -0,0 +1,120 @@
+git checkout master
+
+# Note down the current latest tag because squashing the tito commit into specfile updates
+# will drop it and tito tag needs it further down the build process in order to determine
+# the next (new) tag.
+last_tag="$( git describe --abbrev=0 --tags )"
+
+# sanity check that ensures the latest commits are in the right order.
+#
+# Either the last two commits are the tito tag commit and the specfile updates (where every
+# tito commit is eventually squashed):
+#
+#         .-Other [CARRY] commit
+#        /  .-Other [CARRY] commit
+#       /  /  .-[CARRY] Specfile updates
+#      /  /  /  .-Tito tag commit
+#     /  /  /  /
+# m--c--c--s--t
+#  \           \
+#   \           `-ose/master/HEAD,v3.5.0.23
+#    `-origin/master/HEAD
+#
+# or the last three commits are the tito tag commit, a web console bump that got created
+# during the previous rebase, and the specfile updates:
+#
+#         .-Other [CARRY] commit
+#        /  .-[CARRY] Specfile updates
+#       /  /  .-[DROP] webconsole bump
+#      /  /  /  .-Tito tag commit
+#     /  /  /  /
+# m--c--c--s--t
+#  \           \
+#   \           `-ose/master/HEAD,v3.5.0.23
+#    `-origin/master/HEAD
+#
+# In the first case, we simply want to reset the last two commits and commit them as one.
+# In the second case, we want the same thing as in the first case plus to completely drop
+# the webconsole bump since a new one may be created during the current build.
+sanity_check
+
+# Depending on the state above, we will want to reset either the two or the three latest
+# commits, assuming our sanity check allows us to continue. N will be either 2 or 3.
+git reset --soft HEAD~N
+
+# Squash the tito commit into the specfile updates
+git add .tito origin.spec
+git commit -m "[CARRY][BUILD] Specfile updates"
+
+# We need to retag, because the previous squash removed the latest tito tag.
+git tag "${last_tag}" HEAD
+
+# Rebase OSE on Origin
+git rebase upstream/master
+```
+
+New releases should be cut as they already are today (a mixture of git and tito). CVEs and bugfixes
+should continue to land in release branches as they already are today too. CVEs and important bugfixes
+will continue to be backported back to Origin manually but we may eventually build a script that will
+create git commits locally automatically, eg. hack/cherry-pick-from-ose.sh openshift/ose/pull/578.
+
+Common conflicts can rise when new changes step on branding or tooling code. Manual resolution is the
+only thing we can do in such cases but we can reduce the amount of conflicts by using build flags that
+will replace branding instances depending on the repository that is being built. We should also investigate
+whether we can stop checking in various code like tito diffs, generated docs[1], and the ose-images script.
+Avoiding those diffs will reduce even further the number of conflicts.
+
+In case of manual conflict resolution, or when a new [CARRY] commit needs to be introduced in master, or
```

Comment:
Not clear to me what you are saying here. Where `pull` gets into the picture? Because of the interactive rebase? I think it will be easier to resolve conflicts by using `git rebase -i` rather than resetting and cherry-picking the carries back.